### PR TITLE
Introduce ParagraphBoundary subclass for text editing

### LINKED ISSUE #114180: TextField keyboard shortcut alt + shift + arrow down on MacOS should extend to paragraph boundary
On MacOS native the alt + shift + arrow down keyboard shortcut on a TextField should extend the selection to the paragraph boundary. 

Expected: selection is extended to paragraph boundary.
Actual: selection is extended to line break boundary. Subsequent arrow down events do not extend the selection further past the first selected line.

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
Fixes #114180

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:218]: nit: This does the same thing as `getTextBoundaryAt`. Maybe just implement it there?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:219]: You can search using code points, no? That seems to be easier to implement and doesn't depend on the characters package.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:227]: There're other mandatory line breakers: https://www.unicode.org/standard/reports/tr13/tr13-5.html

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (justinmc) [inline packages/flutter/lib/src/services/text_boundary.dart:186]: "terminator" => "terminators" :robot:

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:196]: nit: consider implementing these 2 separately. `getTextBoundaryAt` always searches for both boundaries instead of the one in the specified direction.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (justinmc) [inline packages/flutter/lib/src/services/text_boundary.dart:212]: Should these comments use three slashes?  I guess as it is right now, users on the docs site will see the docs for TextBoundary.getTextBoundaryAt, and these comments won't show up anywhere.  Is that intentional?

Comment (justinmc) [reply packages/flutter/lib/src/services/text_boundary.dart:212]: Nit: "returning" => "returned"

File: packages/flutter/lib/src/services/text_layout_metrics.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_layout_metrics.dart:66]: When there is a CRLF you shouldn't break after CR

File: packages/flutter/lib/src/services/text_layout_metrics.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_layout_metrics.dart:68]: just return false?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_boundary.dart:212]: "the line terminator" => "both line terminators" or "the line terminator in both directions"

Right, it includes both, one at each end?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (justinmc) [inline packages/flutter/lib/src/services/text_boundary.dart:216]: Nit: Maybe make the language here agnostic of "tap", since this method doesn't necessarily have to do with tapping on the face of it.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (justinmc) [inline packages/flutter/lib/src/services/text_boundary.dart:225]: Nit: use the variable tappedTextOffset instead of position.offset.

File: packages/flutter/test/services/text_boundary_test.dart
Comment (justinmc) [inline packages/flutter/test/services/text_boundary_test.dart:69]: Nit: If you think it would be useful, maybe test one or more edge cases of getTextBoundaryAt, like what happens if you pass it a series of newline characters.

File: packages/flutter/lib/src/widgets/text_editing_intents.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/text_editing_intents.dart:232]: Now that I'm looking at these side by side, it bugs me that "collapseAtReversal" and "continuesAtWrap" use different tenses.  It should probably have been "continueAtWrap".  Not the fault of this PR though :)

**Review by justinmc — APPROVED** [review]
LGTM with nits :+1:

File: packages/flutter/lib/src/services/text_layout_metrics.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_layout_metrics.dart:68]: do you mean true?

File: packages/flutter/lib/src/services/text_layout_metrics.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_layout_metrics.dart:66]: Do you mean the CR should come before the LF in this switch?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (justinmc) [reply packages/flutter/lib/src/services/text_boundary.dart:212]: About my first comment, actually I think the docs site will show both the current docs and inherited docs?  See: https://master-api.flutter.dev/flutter/dart-core/List/first.html

So I guess if you did add triple slashes here, the docs site would show both.  Just letting you know in case you choose to do it.

File: packages/flutter/lib/src/services/text_layout_metrics.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_layout_metrics.dart:66]: huh the unicode website is offline. You should not break after CR if it's followed by a LF.

File: packages/flutter/lib/src/services/text_layout_metrics.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_layout_metrics.dart:68]: yeah sorry.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart]: I'm okay with triple slashes here, but I noticed the other `TextBoundary` subclasses do not have any additional documentation, so maybe I should just leave it as double slash? Do you have any thoughts @LongCatIsLooong ?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:198]: Use indices to iterate since you'll need them anyways? I think that will make the implementation way simpler.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:141]: Would this work:

```dart
if (position < 0) { 
  return null;
} else if (position >= _text.length) { 
  return getLeadingTextBoundaryAt(_text.length - 1);
}

int index = position;
while (index > 0 && !isNewline(codeUnits[index])) { 
  index -= 1;
}
```

You should be able to implement `getTrailingTextBoundaryAt` in a similar fashion, with an additional look ahead step when `index` points to a `CR`.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:141]: Would this work:

```dart
if (position < 0) { 
  return null;
} 
if (position >= _text.length) { 
  return getLeadingTextBoundaryAt(_text.length - 1);
}
assert(_text.isNotEmpty);
int index = position;

if (index > 0 && codeUnits[index] == LF && codeUnits[index - 1] == CR) index -= 2;
if (isNewline(codeUnits[index])) index -= 1;
while (index > 0 && !isNewline(codeUnits[index])) { 
  index -= 1;
}
return max(index, 0);
```

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart:142]: This fails for something like `abcd efg hi\r\n\n\n\n\n\n\n\n\n\n\n\njklmno\npqrstuv`, given `position` 18. The expected output is 18 for the leading boundary, but we get 17 instead because we skip the new line at the initial position.

**Comment by Renzo-Olivares** [comment]
@justinmc @LongCatIsLooong This is ready for a re-review

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart]: Thank you for that code! Helped simplify the logic a bunch.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart:149]: yeah I think the `while` part needs to be fixed (had an off-by-one moment there):
```dart
while (index > 0) { 
  if (isNewline(codeUnits[index])) { 
    return index + 1;
  }
  index -= 1; 
}
return max(index, 0);
```

File: packages/flutter/test/services/text_boundary_test.dart
Comment (justinmc) [inline packages/flutter/test/services/text_boundary_test.dart:149]: Nit: Up to you if you think this is any more clear, but you could turn this test into `group('paragraph boundary', ...` and then make each independent case below into its own small test.

File: packages/flutter/test/services/text_boundary_test.dart
Comment (justinmc) [reply packages/flutter/test/services/text_boundary_test.dart:207]: I think you've updated this to include a bunch of good edge cases now :+1:

**Review by justinmc — APPROVED** [review]
LGTM when Weiyu's comments are addressed :+1:

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart:141]: Wouldn't we be unable to iterate through the text if the given position is not contained within the text? Or in that case do we set the position to the end of the text and start iterating from there?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart:144]: I'm unsure if we should be returning `null` or `0` in the case (I'm leaning towards 0 since that would have `getTextBoundaryAt` return a collapsed TextRange at 0). Do you have any opinions on this?

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart:142]: Eot is also considered a paragraph break I think. The convention is to return the largest index regardless of the bounds, but I don't know if there're significant benefits over returning null when the index is OOB.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart]: I went with `0` for the reason above.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart]: I think if someone where to use `ParagraphBoundary` a stand-alone class with a text of 'hello world'. If they use `100` for the position given to `getLeadingBoundaryAt` and `getTrailingBoundaryAt`, then they would expect null bounds since `position` `100` does not exist within the text.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart]:
```dart
  /// Returns the offset of the closest text boundary before or at the given
  /// `position`, or null if no boundaries can be found.
  ///
  /// The return value, if not null, is usually less than or equal to `position`.
  int? getLeadingTextBoundaryAt(int position);
```
This is the definition so by that definition it should return the smallest offset, "returning null if OOB" currently is not a promise the interface makes, I think it's OK to make that change but to keep it consistent all subclasses would need to change their behavior. But there doesn't seem to be a huge benefit that you gain from switching from one to another?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart]: Oh okay that makes sense. Sorry I was misunderstanding before, I am fine with keeping the definition with how it is. I have updated the implementation to reflect this.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:186]: nit: `position == _text.length - 1` is a special value but probably doesn't need special treatment.

File: packages/flutter/lib/src/services/text_boundary.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/text_boundary.dart:201]: is this right? You've already `+1`ed in the CRLF case, so if there's CRLFLF wouldn't this return the offset after the second LF?

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_boundary.dart:201]: Maybe something like this:
```dart
    while (!TextLayoutMetrics.isLineTerminator(codeUnits[index])) {
        if  (index == codeUnits.length) {
           return index;
         }
        index += 1;
    }
    return index < codeUnits.length - 1 && codeUnits[index] == 0xD && codeUnits[index + 1] == 0xA
             ?  index + 2
             : index + 1;
```

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart:201]: I think that would still work out because when we break out of the `while` loop the `index` would be at `CR` codeunit then, doing a `+1` for the `CRLF` case and a `+1 ` for the line terminator case would leave it after the first `LF`.

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_boundary.dart:201]: I like your code though, it cleans up a few lines. I think the `index == codeUnits.length` should go after we increment the index though.

File: packages/flutter/lib/src/services/text_layout_metrics.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_layout_metrics.dart:66]: Aren't 0x2028 and 0x2029 also supposed to be on this list?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/services/text_layout_metrics.dart:66]: Done, thanks for catching that!

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:3923]: can you just use _moveBeyondTextBoundary instead?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/editable_text.dart]: That worked! I think I was tying myself up here when trying to implement `getLeadingTextBoundaryAt` and `getTrailingTextBoundaryAt`.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:4914]: Could you explain this change? We do care about the affinity here I think?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/editable_text.dart:4914]: I made this change because when  `selection.base` had an `upstream` affinity, for example when we have a word wrapped line and place the caret in the `downstream` position at the word wrap and try to get the leading paragraph boundary, the `selection.extendTo` will create a selection with an `upstream` affinity, so when trying to collapse the selection after, the caret is placed at `upstream` position at the word wrap. It is expected that it would be in the `downstream` position of the word wrap where the selection began. 

I can also confirm this behavior on native macOS. And from my testing the change hasn't affected other shortcuts.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:4914]: Is that the case for word boundary too? If it is could you add a test to make sure this behavior won't get changed?

**Review by LongCatIsLooong — APPROVED** [review]
LGTM once the collapse behavior is confirmed.

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Offline: word boundary on native macOS also behaves this way (defaults to `downstream`), but native Linux and Windows seem to match the behavior where they care about the `affinity`. I'm going to hold off on this `.fromPosition` change since it functionally doesn't break anything, but I will open up an issue to fix this default behavior on macos.

**Pull request merged**
